### PR TITLE
MatrixFree: allow categorization of ghost cells

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: source/matrix_free/task_info.cc [hunk 1]
```diff
@@ -1014,10 +1014,61 @@ namespace internal
           blocks                      = {0, comm_begin, comm_end, end};
         }
 
-      // Step 7: Fill in the data by batches for the locally owned cells.
+      // Step 7: sort ghost cells according to the category
+      std::vector<std::tuple<unsigned int, unsigned int>>
```
Comment (kronbichler) [inline source/matrix_free/task_info.cc:1018]: Is there a reason to choose `tuple` over `std::array<unsigned int, 2>` as is used in line 968 above?

**Review by kronbichler — APPROVED** [review]

File: source/matrix_free/task_info.cc [hunk 2]
```diff
@@ -1014,10 +1014,62 @@ namespace internal
           blocks                      = {0, comm_begin, comm_end, end};
         }
 
-      // Step 7: Fill in the data by batches for the locally owned cells.
+      // Step 7: sort ghost cells according to the category
+      std::vector<std::array<unsigned int, 2>> tight_category_map_ghost;
```
Comment (peterrum) [inline source/matrix_free/task_info.cc:1018]: @kronbichler Done!

File: source/matrix_free/task_info.cc [hunk 3]
```diff
@@ -1014,10 +1014,62 @@ namespace internal
...
+                counters[index]++;
+            }
+
+          // insert padding
+          for (unsigned int i = 0; i < counters.size(); ++i)
+            if (counters[i] % n_lanes != 0)
+              for (unsigned int j = counters[i] % n_lanes; j < n_lanes; ++j)
+                tight_category_map_ghost.emplace_back(
+                  std::array<unsigned int, 2>{
+                    {i, numbers::invalid_unsigned_int}});
```
Comment (kronbichler) [inline source/matrix_free/task_info.cc:1060]: Shouldn't it suffice to write
```suggestion
                    i, numbers::invalid_unsigned_int);
```

Comment (kronbichler) [reply source/matrix_free/task_info.cc:1060]: Sorry, I was wrong, this won't work. We should keep your code.

**Review by kronbichler — APPROVED** [review]

File: source/matrix_free/task_info.cc (hunk 3 above)
Comment (peterrum) [reply source/matrix_free/task_info.cc:1060]: No. This is not working.

**Pull request merged**